### PR TITLE
ath79: Fix builtin-switch support for ar934x

### PR DESCRIPTION
**Pull request by 981213**
Please check commit message on each commit for details. I'm too lazy to type them again :-D
It's only tested AR9344.

To use this on AR934x with phy0 and phy4 swapped:
```
&mdio1 {
	status = "okay";
	phy4-mii-enable;
	phy0: ethernet-phy@0 {
		reg = <0>;
		phy-mode = "mii";
	};
};

&eth1 {
	status = "okay";

	gmac-config {
		device = <&gmac>;
		switch-phy-swap = <1>;
		/* The following line is only needed for AR9344. Remove this if you are using AR9341 */
		switch-only-mode = <1>;
	};
};

&eth0 {
	status = "okay";
	phy-mode = "mii";
	phy-handle = <&phy0>;
	qca,delay-probe;
};

```
To use this on AR934x without phy0 and phy4 swapped:
```
&mdio1 {
	status = "okay";
	phy4-mii-enable;
	phy4: ethernet-phy@4 {
		reg = <4>;
		phy-mode = "mii";
	};
};

&eth1 {
	status = "okay";

	gmac-config {
		device = <&gmac>;
		/* The following line is only needed for AR9344. Remove this if you are using AR9341 */
		/* gmac-config is still needed even if it doesn't set anything because we want to clear bits in ETH_CFG that might be set by U-boot. */
		switch-only-mode = <1>;
	};
};

&eth0 {
	status = "okay";
	phy-mode = "mii";
	phy-handle = <&phy4>;
	qca,delay-probe;
};
```

### CODE REVIEW

File: target/linux/ath79/files/drivers/net/ethernet/atheros/ag71xx/ag71xx_main.c [hunk 1]
```diff
@@ -1342,6 +1366,15 @@ static int ag71xx_probe(struct platform_device *pdev)
 	if (!np)
 		return -ENODEV;
 
+	err = ag71xx_setup_gmac(np);
+	if (err)
+		return err;
+
+	if (of_device_is_compatible(np, "qca,ar9340-eth") && (!ag71xx_gmac_configured)) {
```
Comment (981213) [inline target/linux/ath79/files/drivers/net/ethernet/atheros/ag71xx/ag71xx_main.c]: I didn't realize that this problem exists on other SoCs and I applied this workaround only for ar934x.
@lucize You should add a compatible string to use this workaround.

Comment (lucize) [reply target/linux/ath79/files/drivers/net/ethernet/atheros/ag71xx/ag71xx_main.c]: yes, I meant to ask you about this, I'll try

**Comment by 981213** [comment]
I got confused now: Does the builtin switch work on ar933x/ar724x? I've just noticed that the builtin switch is always connected to mdio1 so mdio1 should always be initialized first. In the old ar71xx target, mdio and gmac are registered as two separated devices and the driver initialize both mdios before initializing gmac0/gmac1. But in ath79 the mdio is initialized during the initialization of gmac so gmac1 should be initialized before gmac0 on all SoCs with a builtin switch.

**Comment by 981213** [comment]
If all devices with 2 gmacs should have the second one initialized first, reordering it in dts should be a better workaround than returning a -EPROBE_DEFER in probe function.

File: target/linux/ath79/files/drivers/net/ethernet/atheros/ag71xx/ag71xx_main.c (hunk 1 above)
Comment (lucize) [reply target/linux/ath79/files/drivers/net/ethernet/atheros/ag71xx/ag71xx_main.c]: I added something like 

`if ((of_device_is_compatible(np, "qca,ar9340-eth") || of_device_is_compatible(np, "qca,qca9530-eth")) && (!ag71xx_gmac_configured)) {`

**Comment by 981213** [comment]
for the RFC:
It seems that we should always get mdio1 initialized before doing anything else.
What the ar71xx did:
1. Configure ETH_CFG for phy swapping etc. in machine setup function.
2. Register the mdio devices needed.
3. Register gmac devices needed.
mdio is a separated platform device in ar71xx so we got two mdios and builtin switch initialized before gmac0/gmac1.

What we do now in ath79:
We configure ETH_CFG/mdio/gmac in the same platform device.
1 Configure ETH_CFG if a gmac-config node present in dts.
2 Initalize gmac.
3 Before polling connection status we initialize mdio corresponding to current gmac.
In this initialization sequence we will have gmac0 configured before mdio1 but when using builtin switch the connection status for gmac0 should be polled from mdio1 which isn't ready.

Here are some ways I could thought of:
1 Return -EPROBE_DEFER like what I've done.
2 Reordering gmac0 and gmac1 in ath79.dtsi which I think isn't a good way because the devices in dts is designed to be ordered by address instead of probe order.
3 Split mdio and builtin switch out of the gmac initialization and probe them as a separate platform device before gmac which I don't know if it's possible.

Questions here:
Does the builtin switch work on ar933x/ar724x with both gmacs?
Which way of reordering device initialization would you recommend?

**Comment by juppin** [comment]
> We can throw gmac0 away and use gmac1 with VLAN. But this make things inefficient.

As long as we cloud not use the wan features of gmac0 i don´t think this would be as much inefficient...
All builtin switches operate only with 100M on the physical port side and does have a 1G connection to the cpu.
I´ve running this setup on my (WR841v8) ar9341 for some time, but only as dump ap.

File: target/linux/ath79/files/drivers/net/ethernet/atheros/ag71xx/ag71xx_main.c (hunk 1 above)
Comment (juppin) [reply target/linux/ath79/files/drivers/net/ethernet/atheros/ag71xx/ag71xx_main.c]: As @lucize tested it on WR841v9, the qca953x soc´s does not need this fix.

Comment (lucize) [reply target/linux/ath79/files/drivers/net/ethernet/atheros/ag71xx/ag71xx_main.c]: with

[    0.625225] libphy: Fixed MDIO Bus: probed
[    1.311983] libphy: mdio: probed
[    1.682981] mdio-bus.0: Found an AR934X built-in switch
[    1.763668] ag71xx 1a000000.eth: connected to PHY at fixed-0:00 [uid=00000000, driver=Generic PHY]
[    1.773753] eth0: Atheros AG71xx at 0xba000000, irq 5, mode:GMII
[    2.481983] libphy: mdio: probed
[    2.486817] ag71xx 19000000.eth: connected to PHY at mdio-bus.0:04 [uid=004dd042, driver=Generic PHY]
[    2.497514] eth1: Atheros AG71xx at 0xb9000000, irq 4, mode:MII

without

[    0.400919] libphy: Fixed MDIO Bus: probed
[    1.083694] libphy: mdio: probed
[    1.087067] ag71xx 19000000.eth: Could not connect to PHY device
[    1.773690] libphy: mdio: probed
[    2.114686] mdio-bus.1: Found an AR934X built-in switch
[    2.169405] ag71xx 1a000000.eth: connected to PHY at fixed-0:00 [uid=00000000, driver=Generic PHY]
[    2.179477] eth0: Atheros AG71xx at 0xba000000, irq 5, mode:GMII

Comment (juppin) [reply target/linux/ath79/files/drivers/net/ethernet/atheros/ag71xx/ag71xx_main.c]: Also if you disable/remove mdio0

Comment (lucize) [reply target/linux/ath79/files/drivers/net/ethernet/atheros/ag71xx/ag71xx_main.c]: if I don't have mdio0 then eth0 or eth1 will not show, one of them, I don't recall, but I tested that also

Comment (lucize) [reply target/linux/ath79/files/drivers/net/ethernet/atheros/ag71xx/ag71xx_main.c]: now is like on ar71xx target, eth1 being the wan, before this patch, eth0 was the wan as port 5 on the switch, now I have only up to port4 on the switch (like ar71xx)

**Comment by lucize** [comment]
@981213  do you think you can have a look at https://github.com/openwrt/openwrt/commit/253095131d7b1a22d9a97cf81640bb3cacc941e4 but that one is using a realtek switch

**Comment by 981213** [comment]
@lucize It doesn't help here. RTL8367RB is configured using GPIO I2C instead of mdio.

**Comment by 981213** [comment]
I made the probe delay activated by a bool value in dts so that this patchset will only apply the probe workaround to ar934x and people can activate it for other SoCs if they need.

**Comment by juppin** [comment]
Was the gmac-config node yesterday also present on wdr4300.dtsi?
Should i do a second test?

**Comment by 981213** [comment]
I've just found out why the builtin switch work on ar933x/ar724x...
```
&mdio0 {
	regmap = <&eth1>;
	builtin-switch;
	resets = <&rst 23>;
	reset-names = "mdio";
};
```
regmap of mdio0 is pointed to eth1. So we have mdio1 and the builtin switch initialized during the initialization of gmac0.
I would prefer my solution because in ag71xx_open function from ath79:
```c
static int ag71xx_open(struct net_device *dev)
{
	struct ag71xx *ag = netdev_priv(dev);
	unsigned int max_frame_len;
	int ret;

	netif_carrier_off(dev);
	max_frame_len = ag71xx_max_frame_len(dev->mtu);
	ag->rx_buf_size = SKB_DATA_ALIGN(max_frame_len + NET_SKB_PAD + NET_IP_ALIGN);

	/* setup max frame length */
	ag71xx_wr(ag, AG71XX_REG_MAC_MFL, max_frame_len);
	ag71xx_hw_set_macaddr(ag, dev->dev_addr);

	ret = ag71xx_hw_enable(ag);
	if (ret)
		goto err;

	ag71xx_ar7240_start(ag);
	phy_start(ag->phy_dev);

	return 0;

err:
	ag71xx_rings_cleanup(ag);
	return ret;
}
```
We made a call to ag71xx_ar7240_start(ag) which in ar71xx we call this only if current gmac is the one connected to builtin switch.
```c
void ag71xx_ar7240_start(struct ag71xx *ag)
{
	struct ar7240sw *as = ag->phy_priv;

	if (!as)
		return;

	ar7240sw_reset(as);

	ar7240_set_addr(as, ag->dev->dev_addr);
	ar7240_hw_apply(&as->swdev);
}
```
Here we passed the mac address of current gmac to the switch and by using the way ar724x/ar933x did we'll apply the mac from gmac0 instead of gmac1 to switch. I don't know whether this have any side-effects so I still prefer my workaround here.

**Comment by 981213** [comment]
@juppin that node is there in the previous patchset. (The only difference is that I added that node in a separated commit yesterday.)

**Comment by 981213** [comment]
This PR is closed because the current ag71xx mixed up several things and I think it's better to split them apart.
The builtin switch in ar724x/ar933x is ar8216 and the one in ar934x and later is ar8229. I guess we can get it work by some simple changes to the existing ar8216 driver rather than a separated switch driver in ag71xx.
Current mdio1 with builtin-switch defined isn't the mdio belonging to gmac1, it's the one in the builtin switch and we don't actually need to use the mdio in switch to poll phy status. (When MDIO_MASTER_EN isn't set to 1 we can directly operate all those phys with mdio from gmac1 instead of the one in the builtin switch.) (Just on theory. I haven't tested this.)
So I'll split gmac and mdio into two separated platform devices and try to replace the switch driver in ag71xx with a patched ar8216 next week.

And for those who wants to get the builtin switch work, you can try to change mdio0 as what ar9330.dtsi did and configure mdio in gmac1 during initialization of gmac0.

**Pull request closed without merging**